File: Data Structures & Algorithms/three-integer-sum/submission-7.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        l = len(nums)

        res = []
      
        for i in range(l):
            left = i + 1
            right = l - 1

            n1 = nums[i]
            s = None
            while left < right:
                s = n1 + nums[left] + nums[right]
                lst = [n1, nums[left], nums[right]]
                if s > 0:
                    right -= 1
                elif s < 0:
                    left += 1
                elif lst not in res:
                    res.append(lst)
                    left += 1
                    right -= 1

                    while nums[left] == nums[left - 1] and left < right:
                        left += 1
                else:
                    break
                    

                    

        return res

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-7.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        
        while (left < right) {
            int m = (left + right) / 2;

            if (nums[m] > nums[right]) left = m + 1;
            else if (nums[m] <= nums[right]) right = m;
        }

        int pivot = left;

        int search = binarySearch(nums, pivot, nums.length - 1, target);

        return search == -1 ? binarySearch(nums, 0, pivot - 1, target) : search;
    }

    public int binarySearch(int[] nums, int left, int right, int target) {
        while (left <= right) {
            int mid = (left + right) / 2;
            final int midValue = nums[mid];

            if (midValue == target) return mid;

            if (midValue < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }
}
